src/inmanta/deploy/state.py

```diff
@@ -234,38 +199,60 @@ def block_provides(self, resources: Set[ResourceIdStr]) -> set[ResourceIdStr]:
+                # Block resource and its provides.
+                self.block_resource(resource_id, self.resources[resource_id], is_transitive=True)
+                result.add(resource_id)
+                todo.extend(provides_view.get(resource_id, set()))
         return result
 
-    def unblock_resource(self, resource: ResourceIdStr) -> None:
+    def mark_as_defined(self, resource: ResourceIdStr, details: ResourceDetails) -> None:
         """
-        Mark the given resource no longer as blocked. Also mark the provides of the given resource as unblocked (transitively)
-        if they become unblocked because of this.
+        Mark the given resource as defined. If the given resource or any of its provides became unblocked, because of this,
+        this method also updates the blocked status.
 
         Must be called under the scheduler lock. This method assumes that all the resources of the model version
         are populated in the resources dictionary.
 
-        :param resource: The resource that has to be unblocked and potentially any of its provides as a consequence of
-                         `resource` being unblocked.
+        :param resource: The resource that became defined.
+        :param details: The details of resource.
         """
 
-        def _unblock(resource: ResourceIdStr) -> None:
-            self.resource_state[resource].blocked = BlockedStatus.NO
+        def update_blocked_status(resource: ResourceIdStr) -> bool:
+            """
+            Check if the deployment of the given resource is still blocked and mark it as unblocked if it is.
+
+            :return: True iff the given resource was unblocked.
+            """
             if self.resource_state[resource].status is ResourceStatus.UNDEFINED:
-                self.resource_state[resource].status = ResourceStatus.HAS_UPDATE
+                # The resource is undefined.
+                return False
+            if any(self.resource_state[r].blocked.is_blocked() for r in self.requires.get(resource, set())):
+                # Resource is blocked, because a dependency is blocked.
+                return False
+            # Unblock resource
+            self.resource_state[resource].blocked = BlockedStatus.NO
             if self.resource_state[resource].status is ResourceStatus.HAS_UPDATE:
                 self.dirty.add(resource)
+            return True
 
-        _unblock(resource)
-        for res in self.requires.get_all_provides_transitively(resource):
-            is_blocked = self.resource_state[res].status is ResourceStatus.UNDEFINED or any(
-                self.resource_state[r].blocked.is_blocked() for r in self.requires.get(res, set())
-            )
-            if not is_blocked:
-                _unblock(res)
+        provides_view: Mapping[ResourceIdStr, Set[ResourceIdStr]] = self.requires.provides_view()
+        self.resources[resource] = details
+        self.resource_state[resource].status = ResourceStatus.HAS_UPDATE
```

Comment:
Iff we follow my suggestion from the other comment (can't link to in-progress review comments so quoting below), I think these two lines should go.

> Why not simply add these resources to new_desired_state as well? It's semantically accurate and it prevents having a third place where we update resource intent.